components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/context/OptimizedSessionContext.java

```diff
@@ -0,0 +1,193 @@
+
+    private final Map<String, OptimizedSequenceConfig> optimizedAuthenticatedSequences;
+    private final Map<String, OptimizedAuthenticatedIdPData> optimizedAuthenticatedIdPs;
+    private final String tenantDomain;
+    private final boolean isRememberMe;
+    private final Map<String, Object> properties;
+    private final SessionAuthHistory sessionAuthHistory;
+    private final Map<String, Map<String, OptimizedAuthenticatedIdPData>> optimizedAuthenticatedIdPsOfApp;
+
+    private static final Log log = LogFactory.getLog(OptimizedSessionContext.class);
+
+    public OptimizedSessionContext(SessionContext sessionContext) throws SessionDataStorageOptimizationException {
+
+        Object authUser = sessionContext.getProperty(FrameworkConstants.AUTHENTICATED_USER);
+        this.tenantDomain = authUser instanceof AuthenticatedUser ?
+                ((AuthenticatedUser) authUser).getTenantDomain() : null;
+        this.optimizedAuthenticatedSequences = getOptimizedAuthenticatedSequences(
+                sessionContext.getAuthenticatedSequences());
+        this.optimizedAuthenticatedIdPs = getOptimizedAuthenticatedIdPs(sessionContext.getAuthenticatedIdPs());
+        this.isRememberMe = sessionContext.isRememberMe();
+        this.properties = sessionContext.getProperties();
+        this.sessionAuthHistory = sessionContext.getSessionAuthHistory();
+        this.optimizedAuthenticatedIdPsOfApp = getOptimizedAuthenticatedIdPsOfApp(sessionContext.
+                getAuthenticatedIdPsOfApp());
+        if (log.isDebugEnabled()) {
+            log.debug("Optimization process for the session context is completed.");
+        }
+    }
+
+    private Map<String, OptimizedSequenceConfig> getOptimizedAuthenticatedSequences
+            (Map<String, SequenceConfig> authenticatedSequences) throws SessionDataStorageOptimizationException {
+
+        Map<String, OptimizedSequenceConfig> optimizedAuthenticatedSequences = new HashMap<>();
+        for (Map.Entry<String, SequenceConfig> entry : authenticatedSequences.entrySet()) {
+            SequenceConfig sequenceConfig = entry.getValue();
+            for (Map.Entry<Integer, StepConfig> mapEntry : sequenceConfig.getStepMap().entrySet()) {
+                StepConfig stepConfig = mapEntry.getValue();
+                List<AuthenticatorConfig> authenticatorList = stepConfig.getAuthenticatorList();
+                for (AuthenticatorConfig authenticatorConfig : authenticatorList) {
+                    List<String> idPResourceId = new ArrayList<>();
+                    for (Map.Entry<String, IdentityProvider> idpEntry : authenticatorConfig.getIdps().entrySet()) {
+                        IdentityProvider idp = idpEntry.getValue();
+                        String idpName = idpEntry.getKey();
+                        if (StringUtils.isEmpty(idp.getResourceId())) {
+                            idPResourceId.add(getIdPByIdPName(idpName, this.tenantDomain).getResourceId());
```

Comment:
here the tenantDomain should be taken from `idp`

Reply:
`idp` does not have tenantDomain info. So used the SP tenantDomain as discussed.
Resolved with [`8ae0557`](https://github.com/wso2/carbon-identity-framework/pull/4446/commits/8ae05575266f68d589e548a7dc82bf625e079337)